exporter/exporter_test.go

```diff
@@ -0,0 +1,485 @@
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/containerd/containerd/v2/core/content"
+	contentproxy "github.com/containerd/containerd/v2/core/content/proxy"
+	"github.com/containerd/platforms"
+	"github.com/moby/buildkit/client"
+	"github.com/moby/buildkit/client/llb"
+	"github.com/moby/buildkit/exporter/containerimage/exptypes"
+	"github.com/moby/buildkit/frontend/dockerui"
+	gateway "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/moby/buildkit/util/testutil/integration"
+	"github.com/moby/buildkit/util/testutil/workers"
+	digest "github.com/opencontainers/go-digest"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/stretchr/testify/require"
+	"github.com/tonistiigi/fsutil"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+)
+
+func init() {
+	if workers.IsTestDockerd() {
+		workers.InitDockerdWorker()
+	} else {
+		workers.InitOCIWorker()
+		workers.InitContainerdWorker()
+	}
+}
+
+func TestFrontendIntegration(t *testing.T) {
+	testIntegration(t,
+		testGatewayExporter,
+		testIsolatedRemotes,
+		testExternalExporter,
+		testExternalExporterMultiplatform,
+	)
+}
+
+func testIntegration(t *testing.T, funcs ...func(t *testing.T, sb integration.Sandbox)) {
+	// XXX: to build custom exporters we need base images to be available
+	mirroredImages := integration.OfficialImages("golang:1.25-alpine3.22")
+	mirroredImages["tonistiigi/xx:1.6.1"] = "docker.io/tonistiigi/xx:1.6.1"
+	mirrors := integration.WithMirroredImages(mirroredImages)
+
+	tests := integration.TestFuncs(funcs...)
+	integration.Run(t, tests, mirrors)
+}
+
+func testGatewayExporter(t *testing.T, sb integration.Sandbox) {
+	workers.CheckFeatureCompat(t, sb, workers.FeatureOCIExporter, workers.FeatureOCILayout)
+
+	c, err := client.New(sb.Context(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	frontend := func(ctx context.Context, c gateway.Client) (*gateway.Result, error) {
+		st := llb.Scratch().
+			File(llb.Mkfile("/foo.txt", 0644, []byte("foo"))).
+			File(llb.Mkfile("/bar.txt", 0644, []byte("bar")))
+		def, err := st.Marshal(sb.Context())
+		if err != nil {
+			return nil, err
+		}
+		return c.Solve(ctx, gateway.SolveRequest{
+			Definition: def.ToPB(),
+		})
+	}
+
+	var foundFiles []string
+	var foundDescs []ocispecs.Descriptor
+	export := func(ctx context.Context, c gateway.Client, conn *grpc.ClientConn, _ exptypes.ExporterTarget, result *gateway.Result) error {
+		entries, err := result.Ref.ReadDir(ctx, gateway.ReadDirRequest{Path: "/"})
+		if err != nil {
+			return err
+		}
+		for _, entry := range entries {
+			foundFiles = append(foundFiles, entry.Path)
+		}
+
+		store := contentproxy.NewContentStore(conn)
+		descs, err := result.Ref.GetRemote(ctx)
+		if err != nil {
+			return err
+		}
+		foundDescs = filterAvailableDescriptors(ctx, store, descs)
+
+		return nil
+	}
+
+	_, err = c.BuildExport(sb.Context(), client.SolveOpt{}, "", frontend, export, nil)
+	require.NoError(t, err)
```

Comment:
This feels like a fun interface to me - allowing the client to easily do its own local export. This actually works *entirely* without any server-side changes - it's just calling the exporter function right after a build, using the same gateway.

We have access to the entire buildkit content store (so there's no isolation here, but this is client-trusted code).